metatomic-torch/src/io/npy.cpp

```diff
@@ -0,0 +1,373 @@
+#include "metatomic/torch/io/npy.hpp"
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+#include <cstring>
+#include <limits>
+#include <sstream>
+#include <iomanip>
+#include <cctype> // std::isdigit
+
+namespace metatomic_torch {
+namespace io {
+
+// ---- helpers
+
+static inline bool is_ascii(const uint8_t* p, size_t n) {
+  for (size_t i = 0; i < n; ++i) {
+    if (p[i] > 0x7F) return false;
+  }
+  return true;
+}
+
+struct Header {
+  std::string descr;           // '<f8', '<i8', '<i4', '|b1'
+  bool fortran_order;          // must be false
+  std::vector<int64_t> shape;  // non-negative sizes
+};
+
+// Very small, permissive parser for the dict literal we emit.
+// Assumes ASCII input. Tolerates arbitrary spaces and trailing commas like NumPy.
+static Header parse_header_ascii(const std::string& s) {
+  auto eat_space = [&](size_t& i) {
```

Comment:
```suggestion
  auto eat_spaces = [&](size_t& i) {
```

or  `skip_whitespaces`